File: java/src/main/java/tjchiang/leetcode/L0003_Longest_Substring_Without_Repeating_Characters/LongestSubstring.java
package tjchiang.leetcode.L0003_Longest_Substring_Without_Repeating_Characters;

import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;

public class LongestSubstring {

    /**
     * Time complexity: O(n)
     * </br>
     * Space complexity: O(1)
     */
    public int solveBySet(String s) {
        int result = 0;
        int left = 0;
        Set<Character> charSet = new HashSet<>();

        for (int right = 0; right < s.length(); right++) {
            char c = s.charAt(right);
            while (charSet.contains(c)) {
                charSet.remove(s.charAt(left));
                left++;
            }

            charSet.add(c);
            result = Math.max(result, right - left + 1);
        }

        return result;
    }

    /**
     * Time complexity: O(n)
     * </br>
     * Space complexity: O(1)
     */
    public int solveByLastIndex(String s) {
        int result = 0;
        int left = 0;
        Map<Character, Integer> lastIndex = new HashMap<>();

        for (int right = 0; right < s.length(); right++) {
            char c = s.charAt(right);
            if (lastIndex.containsKey(c) && lastIndex.get(c) >= left) {
                left = lastIndex.get(c);
                left++;
            }

            lastIndex.put(c, right);
            result = Math.max(result, right - left + 1);
        }

        return result;
    }
}

File: leetcode/0003-Longest_Substring_Without_Repeating_Characters/longest_substring.py
class LongestSubstring:

    # Time: O(n)
    # Space: O(1)
    def solve_by_set(self, s: str) -> int:
        result = 0
        left = 0
        char_set = set()

        for right in range(len(s)):
            while s[right] in char_set:
                char_set.remove(s[left])
                left += 1

            char_set.add(s[right])
            result = max(result, right - left + 1)

        return result

    # Time: O(n)
    # Space: O(1)
    def solve_by_last_index(self, s: str) -> int:
        result = 0
        left = 0
        last_index = {}

        for right, c in enumerate(s):
            if c in last_index and last_index[c] >= left:
                left = last_index[c]
                left += 1

            result = max(result, right - left + 1)
            last_index[c] = right

        return result
